src/man/man3/World.3
.TH World 3I "7 Mar 1989" "InterViews" "InterViews Reference Manual"
.SH NAME
World \- control a display
.SH SYNOPSIS
.B #include <InterViews/world.h>
.SH DESCRIPTION
A world was an application's connection to a particular display.
It is currently provided for backward compatibility.
.SH PUBLIC OPERATIONS
.TP
.B "World("
.ns
.TP
.B "    const char* classname, int& argc, char** argv"
.ns
.TP
.B "    OptionDesc* = nil, PropertyData* = nil"
.ns
.TP
.B ")"
.br
Construct the world object for a display.
The world's instance name comes from the \-name option if
it was given; otherwise the instance name comes from the environment
variable RESOURCE_NAME if it is nonnil.  If the name has still not
been defined, then argv[0] with all leading directories stripped is
used.  The following arguments are interpreted automatically and are
removed from \fIargv\fP:
.PP
.RS
.nf
.ta 1.5i
\-background	next argument sets the background color
\-bg	same as \-background
\-display	next argument specifies the target workstation display
\-foreground	next argument sets the foreground color
\-fg	same as \-foreground
\-fn	same as \-font
\-font	next argument sets the text font
\-geometry	next argument sets the first top-level interactor's position and size
\-iconic	starts up the first top-level interactor in iconic form
\-name	next argument sets the instance name of all top-level interactors
	that don't have their own instance names
\-reverse	swaps default foreground and background colors
\-rv	same as \-reverse
\-synchronous	force synchronous operation with the window system
\-title	next argument sets the first top-level interactor's title bar name
\-xrm	next argument sets an ``attribute: value'' property
.fi
.RE
The geometry specification has the form ``=WxH+XOFF+YOFF''.  A
negative XOFF (YOFF) specifies the offset of the interactor's right
(bottom) edge from the right (bottom) side of the screen.  The
constructor sets \fIargc\fP to the number of uninterpreted arguments
that remain.
.PP
If non-nil, the OptionDesc* parameter is an array of option descriptors
used to parse application-specific attributes.
OptionDesc contains four fields:
a \fIname\fP used on the command line,
a \fIpath\fP specifying the attribute,
a \fIstyle\fP specifying where the value is, and
a default \fIvalue\fP.
Valid styles are \fIOptionPropertyNext\fP (use next argument as an
attribute:value pair, not just the value),
\fIOptionValueNext\fP (use next argument as value),
\fIOptionValueImplicit\fP (use default value),
\fIOptionValueIsArg\fP (use argument as value), and
\fIOptionValueAfter\fP (use remainder of argument as value).
If non-nil, the PropertyData* parameter is an array of structures
that each contain three string fields:
a \fIpath\fP specifying the attribute,
a \fIvalue\fP specifying the value, and
a \fItype\fP specifying the type name.
Attributes are entered in the following order:
first any application defaults (specified by the PropertyData array),
then application defaults from /usr/lib/X11/app-defaults/\fIclassname\fP,
then user defaults (usually in $HOME/.Xdefaults), and finally
command-line arguments.
Thus, command-line arguments override user defaults, and
both override application defaults.
.TP
.B "const char* name()"
Return the instance name associated with the world.
.TP
.B "const char* classname()"
Return the class name associated with the world.
.TP
.B "int argc()"
Return the number of arguments passed to the world
(not counting those that were interpreted by the constructor).
.TP
.B "char** argv()"
Return the argument vector passed to the world.
.TP
.B "Font* font()"
Return the default font associated with the display.
If a value for ``font'' is defined at the top-level of the property sheet,
then it is used.
Otherwise, a system default is used.
.TP
.B "Color* foreground()"
Return the default foreground color associated with the display.
If a value for ``foreground'' is defined
at the top-level of the property sheet,
then it is used.
Otherwise, a system default is used.
.TP
.B "Color* background()"
Return the default background color associated with the display.
If a value for ``background'' is defined
at the top-level of the property sheet,
then it is used.
Otherwise, a system default is used.
.TP
.B "boolean shaped_windows()"
Return whether the display supports non-rectangular windows.
.TP
.B "boolean double_buffered()"
Return whether windows on the display should by default be double-buffered.
.TP
.B "virtual void flush()"
Repair all damaged windows on the display and then send any pending requests
to the window system.
.TP
.B "virtual void sync()"
Repair all damaged windows on the display, send any pending requests
to the window system, and wait for an acknowledgement from the window
system.
.TP
.B "Coord width()"
Return the width in coordinates of the current screen
associated with the display.
.TP
.B "Coord height()"
Return the height in coordinates of the current screen
associated with the display.
.TP
.B "virtual void run()"
Read events from the display, handling them as they arrive, and
stopping when quit is called.
.TP
.B "virtual void quit()"
Terminate the run loop.
.TP
.B "virtual boolean pending()"
Return whether any events are waiting to be read.
.TP
.B "virtual void read(Event&)"
Read the next event from the world associated with the event.
.TP
.B "virtual boolean read(long sec, long usec, Event&)"
Read the next event from the world associated with the event,
but do not wait more than \fIsec\fP seconds and \fIusec\fP microseconds.
Return whether an event was found in the given time.
.TP
.B "virtual void unread(Event&)"
Put the event back on the input queue for the world associated
with the event.
.TP
.B "virtual void poll(Event&)"
Set the event to an artificial motion event based on the current pointer
position and the state of the buttons and meta-keys.
.B "virtual void SetScreen(int s)"
Set the current screen to \fIs\fP.
.TP
.B "static World* current()"
Return the current world.  The current world is set
when a world is created or an event is received for the display
associated with a world.


CHANGES
May 12th 2012  ivtools-1.2.10

Many practical and small extensions to comterp, the scripting language embedded in
comdraw, flipbook, graphdraw, and drawserv.

Dropped complete support for gcc 3.0 and earlier. Things still might work but you'd be better off
upgradeing to gcc 4.0 or staying with ivtools-1.2.9.

Feb. 23rd, 2010  ivtools-1.2.9

- Debian related changes suggested by Agustin Martin.

- usec timeout on comdraw pause command -- allows for writing loops
  that can be exited.

- option for an end-of-line token introduced to LexScan.

- port to 64 bit Fedora

- fix problem with negative numbers in attribute editor

- make File/Save and File/Revert always work regardless of modified flag

Nov. 21st, 2009  ivtools-1.2.8

- significant work on keeping up with gcc and libstdc++.  A wrapper
  for stdio.h was inadvertently getting pulled in by /usr/include
  files, causing a lot of trouble.  Now builds with gcc-4.4 on Ubuntu
  9.10.

August 25th, 2009  ivtools-1.2.7

- series of memory leak checking commands: alistleak, attrvleak, commandleak, compleak, graphlicleak, mlineleak, viewleak.  Each returns the current number of allocated objects of that type (AttributeValueList, AttributeValue, Command, OverlayComp, Graphic, MultiLineObj, and OverlayView).

- added reference counting to OverlayView's (ComponentView's) referred from AttributeValue's (#define RESOURCE_COMPVIEW)

- fixed attrname func.

- fixed stddev func.

- added BooleanType to NumFunc promotion method.

- fixed graphdraw copy/paste of graph fragments

- new SocketObj for use comterp

- fix behavior of symbol manipulating funcs, by using internal bquote.

- port to Ubuntu Heron

- applied (most of) patch to build on Fedora 10 contributed by John Heidemann

- balance of changes to support ipl-1.0.4

July 9th, 2008  ivtools-1.2.6

- Debian contributed patch for iostreams backward compatibility

- bug fix for dispatcher.c contributed by Damon Permezel

- add mute command to comterp

- balance of changes to support ipl-1.0.2

March 4th, 2008  ivtools-1.2.5

- overall evolution to support use by ipl-1.0.1 (see
  http://sf.net/projects/ipl for further details)

July 20th, 2006  ivtools-1.2.4

- changes for compiling with gcc-4.0

- changes to socket-driven (ACE) use of comterp to support multi-line commands

- changes to allow re-use of scanner/parser with multiple operator tables

October 7th, 2005  ivtools-1.2.3

- add support for an external libtiff with a --with-tiff (and
-with-tiff-lib) configure arguments.

- changes necessary to compile with gcc-4.0

- changes necessary to compile with gcc-4.0 on top of ACE-5.4.7

- changes necessary to compile on suse-9.3 linux (amd64 platform).
Contributed by Paul Vixie.

- fixed bug in "addtool" command/mechanism in graphdraw and drawserv

- added support for color expressions to graphdraw (:colexpr), to
allow attribute-sensitive computation of a node's color.

January 25th, 2005  ivtools-1.2.2

If you submitted a patch in the past half year to ivtools that you
don't see reflected here, please resubmit.  I lost that e-mail.

- fix for problem writing to /tmp

- compilation fix on NetBSD for comterp.c

- compilation patch submitted by Michal Palczew.

July 13th, 2004  ivtools-1.2.1

Two crashes discovered and fixed by Michal Palczewski (mpalczew at
users.sourceforge.net).  One was a crash on exit after a delete with
idraw/drawtool, the other was a crash upon delete with graphdraw.
They turned out to be long-standing bugs in the order of object
deletion, problems that hadn't appeared with other versions of gcc.
RedHat must have adopted a more aggressive recycling of freed space in
a recent release.  Thanks Michal.


June 3rd, 2004  ivtools-1.2

ivtools-1.2 sets a new direction for the package, toward peer-to-peer
drawing chat.  ivtools' drawserv, the editor that incorporates every ivtool
mechanism to date (multi-frame, graph-editing, integrated scripting
language) has been extended to support a unique kind of
small-granularity shared whiteboard.  

This work was based on the pre-existing networked scripting capability
of ivtools' comdraw.  Then the pasting of any graphic was intercepted,
and the graphic serialized into a command that could be communicated
to other drawing editors to keep them in sync.

Then a connection management system was established, that supported
two-way peer-to-peer links between drawing editors.  In addition the
return path on each socket was used for detecting timeout, which
causes the connection to be dropped.

Finally, a distributed mutually-exclusive graphic selection mechanism
was created, where every drawing editor is aware of who currently has
a graphic selected or who had it selected last if it is not currently
selected.  When one drawing editor tries to select a graphic, it first
requests that permission from the last selector (after determining it
is unselected).

Now you can link to other drawing editors, draw graphics and see them
appear remotely (although only in black and white), and exercise the
distributed selection mechanism.  Four pieces remain to be developed:
1) support for full-up graphic state transmittal (it is sent, but not
interpreted on the remote end);, 2) support for transmission of the
complete set of drawing editor commands; 3) extending the distributed
mutual-exclusion mechanism to select pseudo-graphics to acquire global
permission for connection, pasting, moving front-to-back, and
grouping; and 4) a two-tier mechanism for invoking select(), that will
allow the bypass of selection requests while engaged in lengthy direct
manipulations.

If all this interests you, and you'd like to help, build this version
of ivtools with ACE support, subscribe to the ivtools-dev mailing, and
help me do some testing, programming, thinking, whatever.  Or visit
http://www.ivtools.org to make a donation to support this work.  Thanks.

Scott Johnston




src/IV-2_6/world.cc
/*
 * Copyright (c) 1987, 1988, 1989, 1990, 1991 Stanford University
 * Copyright (c) 1991 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Stanford and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Stanford and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 *
 * IN NO EVENT SHALL STANFORD OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */

/*
 * A world is the root scene for a display.
 */

#include <InterViews/display.h>
#include <InterViews/color.h>
#include <InterViews/font.h>
#include <InterViews/session.h>
#include <InterViews/style.h>
#include <IV-2_6/InterViews/sensor.h>
#include <IV-2_6/InterViews/world.h>
#include <OS/string.h>

#ifndef PDP_PLUS_PLUS
double inch, inches, cm, mm, point, points;
#else
double ivinch, ivinches, ivcm, ivmm, ivpoint, ivpoints;
#endif

World* World::current_;

World::World() {
    session_ = nil;
    display_ = nil;
}

World::World(
    const char* classname, int& argc, char** argv,
    const OptionDesc* opts, const PropertyData* initprops
) {
    session_ = new Session(classname, argc, argv, opts, initprops);
    display_ = session_->default_display();
    make_current();
    Sensor::init();
}

World::~World() {
    delete session_;
}

Session* World::session() const { return session_; }
Display* World::display() const { return display_; }
const char* World::name() const { return session_->name(); }
const char* World::classname() const { return session_->classname(); }
int World::argc() const { return session_->argc(); }
char** World::argv() const { return session_->argv(); }

Style* World::style() const {
    return display_->style();
}

const char* World::property_value(const char* s) const {
    String v;
    if (style()->find_attribute(String(s), v)) {
	return v.string();
    }
    return nil;
}

boolean World::property_is_on(const char* s) const {
    return style()->value_is_on(String(s));
}

const Font* World::font() const {
    const Font* f = nil;
    Style* s = style();
    String v;
    if (s->find_attribute("font", v) || s->find_attribute("Font", v)) {
	f = Font::lookup(v);
    }
    if (f == nil) {
	f = Font::lookup("fixed");
    }
    return f;
}

const Color* World::foreground() const {
    const Color* c = nil;
    Style* s = style();
    String v;
    if (s->find_attribute("foreground", v) ||
	s->find_attribute("Foreground", v)
    ) {
	c = Color::lookup(display(), v);
    }
    if (c == nil) {
	c = new Color(0.0, 0.0, 0.0, 1.0);
    }
    return c;
}

const Color* World::background() const {
    const Color* c = nil;
    Style* s = style();
    String v;
    if (s->find_attribute("background", v) ||
	s->find_attribute("Background", v)
    ) {
	c = Color::lookup(display(), v);
    }
    if (c == nil) {
	c = new Color(1.0, 1.0, 1.0, 1.0);
    }
    return c;
}

boolean World::shaped_windows() const {
    return style()->value_is_on("shaped_windows");
}

boolean World::double_buffered() const {
    return style()->value_is_on("double_buffered");
}

void World::flush() { display_->flush(); }
void World::sync() { display_->sync(); }
Coord World::width() const { return display_->width(); }
Coord World::height() const { return display_->height(); }
unsigned int World::pwidth() const { return display_->pwidth(); }
unsigned int World::pheight() const { return display_->pheight(); }

void World::run() { session_->run(); }
void World::quit() { session_->quit(); }
boolean World::done() const { return session_->done(); }

boolean World::pending() const { return session_->pending(); }
void World::read(Event& e) { session_->read(e); }
boolean World::read(long sec, long usec, Event& e) {
    return session_->read(sec, usec, e);
}
void World::unread(Event& e) { session_->unread(e); }
void World::poll(Event& e) { session_->poll(e); }

void World::RingBell(int i) { display_->ring_bell(i); }
void World::SetKeyClick(int i) { display_->set_key_click(i); }
void World::SetAutoRepeat(boolean b) { display_->set_auto_repeat(b); }
void World::SetFeedback(int t, int s) { display_->set_pointer_feedback(t, s); }

void World::SetScreen(int s) { display_->set_screen(s); }

World* World::current() {
    if (current_ == nil) {
	World* w = new World;
	w->session_ = Session::instance();
	w->display_ = w->session_->default_display();
	w->make_current();
	Sensor::init();
    }
    return current_;
}

void World::make_current() {
    current_ = this;
    ivpoint = double(display_->pwidth()) / display_->a_width();
    ivpoints = ivpoint;
    ivinch = 72.27 * ivpoint;
    ivinches = ivinch;
    ivcm = ivinch / 2.54;
    ivmm = ivinch / 25.4;
}
